File: src/main/java/orbitsim/StateVector.java
package orbitsim;

/** ECI 위치 [km] · 속도 [km/s]. */
public record StateVector(Vector3 positionKm, Vector3 velocityKmS) {

    /** 비에너지 ε = v²/2 − μ/r [km²/s²]. 이체 문제에서 보존된다. */
    public double specificEnergy() {
        double v = velocityKmS.norm();
        double r = positionKm.norm();
        return v * v / 2.0 - Constants.MU_EARTH / r;
    }

    /** 비각운동량 h = r × v [km²/s]. 이체 문제에서 보존된다. */
    public Vector3 specificAngularMomentum() {
        return positionKm.cross(velocityKmS);
    }
}

File: csharp/OrbitSim/StateVector.cs
namespace OrbitSim;

/// <summary>ECI 위치 [km] · 속도 [km/s].</summary>
public readonly record struct StateVector(Vector3 PositionKm, Vector3 VelocityKmS)
{
    /// <summary>비에너지 ε = v²/2 − μ/r [km²/s²]. 이체 문제에서 보존된다.</summary>
    public double SpecificEnergy()
    {
        double v = VelocityKmS.Norm();
        double r = PositionKm.Norm();
        return (v * v / 2.0) - (Constants.MuEarth / r);
    }

    /// <summary>비각운동량 h = r × v [km²/s]. 이체 문제에서 보존된다.</summary>
    public Vector3 SpecificAngularMomentum() => PositionKm.Cross(VelocityKmS);
}
